### ISSUE TITLE
As a user, I would like to receive DMs in Slack when a PR gets a merge conflict

### ISSUE THREAD

**Issue by jchavarri** [pullreminders/backlog#42]
There is a field in the pull request object called `mergeable` that allows to know if a PR can be merged: https://developer.github.com/v3/git/#checking-mergeability-of-pull-requests

It would be very valuable as a user to receive a notification when a pull request goes from `mergeable: true` to `mergeable: false`.

**Comment by abinoda** [comment]
@jchavarri Thanks for submitting this! Could you elaborate on the use case a bit more? Are you hoping to receive a notification when one of your pull requests gets the [required number of approvals](https://help.github.com/articles/enabling-required-reviews-for-pull-requests/)? If so I don't think the `mergeable` property is the best way to do it because I believe that property also depends on whether there are Git conflicts or not.

**Comment by jchavarri** [comment]
Thanks for the quick answer! Our use case is as follows: we have a process where when a PR is ready, we assign it for code review first, and then it goes through a second QA step, where the PR gets tested.

During these steps, chances are that the PR gets in conflict with the main development branch, and these conflicts need to be resolved and in general involve repeating some of these steps.

By having a notification of the `mergeable` state as soon as it changes to "unmergeable", I would be able to know instantly if one of my PRs have merge conflicts, and this would allow me to solve them quickly, reducing the chance that these situations occur.

Does that make sense? Let me know if you want me to elaborate. Thanks!

**Comment by abinoda** [comment]
@jchavarri This makes a lot of sense. Will move it up in the backlog and hopefully get to it pretty soon!

**Comment by abinoda** [comment]
This was just +1'd by @jamesdh:

> Sending reminders (maybe instant "notifications" as well) when any PR's you have created have merge conflicts.

**Comment by abinoda** [comment]
Technical approach as discussed with GitHub team:

> Use my existing subscription to pull_request webhook events to identify when a PR is merged (via closed event). Use this as the "triggering event" to queue a job that executes a minute or two later (to allow time for GitHub's background job to complete) that checks other open PRs (using conditional requests) in the same repo for changes in mergeable.

**Comment by abinoda** [comment]
I began work on this feature today but ran into some issues. After doing some more testing and research, I believe this feature is unfeasible given the current capabilities of GitHub's API.

What I am trying to build is a Slack DM notification for pull request authors when their pull requests go into a state of having a merge conflict, either from new code that was pushed or the base branch being updated (ie. another pull request being merged). Outside of Pull Reminders, merge conflict email notifications have been requested and discussed here https://github.com/isaacs/github/issues/224. It's also worth mentioning that [GitLab natively offers merge conflict email notifications](https://docs.gitlab.com/ee/workflow/notifications.html).

Since there's no GitHub webhook event for "change in mergeability status", I need to detect changes in status myself by keeping a cache of the current mergeability statuses which I can compare with changes after new code is pushed to a pull request or the base branch is updated (potentially creating merge conflicts for all pull requests targeting this branch).

GitHub does not calculate mergeability *unless the pull request is loaded via the UI or an API request*. Here's [GitHub's documentation on this](https://developer.github.com/v3/git/#checking-mergeability-of-pull-requests). As should be expected then, I'm seeing that most pull requests currently cached in Pull Reminders have a `null` value for `mergeable` since on the first API request for a pull request, GitHub begins calculating mergeability and returns `null`. To get the actual value, I would need to make an additional API request shortly thereafter (ie. after 30 seconds) or do some polling to get the value once GitHub has finished calculating.

The problem with what I've just described is that it means this feature would require *a lot* of API requests and easily burn through GitHub's rate limits.

In order to detect changes in pull request mergeability, every time there is a `synchronized` (push) webhook event for a pull request, I would need to make *two or more* API requests to obtain the mergeable state of a pull request – one to trigger the mergeability calculation, and then one or more additional requests to get the value. When there's an update to a base branch (ie. `master`), I would need to loop through all open pull requests with that base branch and check their mergeability statuses with two or more API requests.

This means 2 API requests per push, and `2 * number of open PRs` API requests whenever a base branch is updated (ie. when a pull request is merged). Take for example an organization with 100 engineers where each user is pushing 10 times per hour, 2 PRs are merged per hour, and there are 100 PRs open across the org at a time. This would result in 6,000 API requests per hour (see calculation below). This would technically exceed the rate limit, but with slightly different numbers it would. And this doesn't factor in other API requests that Pull Reminders makes.

```
100 * 10 * 2      = 2000 API req/hour for checking PR mergeability after code is pushed
100 * 10 * 2 * 2  = 4000 API req/hour for checking PR mergeability after base ref updated
```

I am actively discussing this with GitHub Support and exploring options. Merge conflicts are a common event in a developer's day-to-day workflow so I'd hope we can find a way to support this.

Some ideas off the top of my head for resolving this would be:

* GitHub providing a special rate limit increase for Pull Reminders in order to be able to support this feature
* Github adding a webhook event for changes to mergeability that fires without a pull request being explicitly requested
* GitHub calculating mergeability automatically without the PR having to be requested first – this cuts down the number of API requests needed by at least half. I would simply delay the check for mergeability status for a minute or so after a triggering event (push or base branch update).

**Comment by abinoda** [comment]
Update: @swinton suggested seeing if conditional requests could be used to trigger GitHub into calculating mergeability without eating into the rate limit. Turns out, it can 🎉 This reduces the number of API requests required by half (compared to my original projections). I'm not sure if this completely solves my concerns around rate limits but it does certainly help. I'm going to run some more numbers and see if it seems safe enough to move forward.

**Comment by abinoda** [comment]
Update: I did some numbers and am unblocked on this. Hope to get this out this week!

**Comment by abinoda** [comment]
This has been released in closed beta. @jchavarri I've enabled it for your team. Please try it out and let me know how it works!

**Comment by usn1485** [comment]
Hi @abinoda what do I need to do  for github to direct message me on merge conflicts?

**Comment by dsacramone** [comment]
> Hi @abinoda what do I need to do for github to direct message me on merge conflicts?

You ever sort this out, if so- how?

**Comment by codemasta** [comment]
Please is there a solution for this now on Github ?
